libs/geo/src/main/java/org/elasticsearch/geometry/utils/SpatialEnvelopeVisitor.java

```diff
@@ -0,0 +1,302 @@
+            maxX = Math.max(maxX, x);
+            maxY = Math.max(maxY, y);
+        }
+
+        @Override
+        public void visitRectangle(double minX, double maxX, double maxY, double minY) {
+            if (minX > maxX) {
+                throw new IllegalArgumentException("Invalid cartesian rectangle: minX > maxX");
+            }
+            this.minX = Math.min(this.minX, minX);
+            this.minY = Math.min(this.minY, minY);
+            this.maxX = Math.max(this.maxX, maxX);
+            this.maxY = Math.max(this.maxY, maxY);
+        }
+
+        @Override
+        public boolean isValid() {
+            return minY != Double.POSITIVE_INFINITY;
+        }
+
+        @Override
+        public Rectangle getResult() {
+            return new Rectangle(minX, maxX, maxY, minY);
+        }
+    }
+
+    /**
+     * The geographic point visitor determines the envelope by the minimum and maximum x/y coordinates,
+     * while allowing for wrapping the longitude around the dateline.
+     * When longitude wrapping is enabled, the visitor will determine the smallest bounding box between the two choices:
+     * <ul>
+     *     <li>Wrapping around the front of the earth, in which case the result will have minx &lt; maxx</li>
+     *     <li>Wrapping around the back of the earth, crossing the dateline, in which case the result will have minx &gt; maxx</li>
+     * </ul>
+     */
+    public static class GeoPointVisitor implements PointVisitor {
+        private double minY = Double.POSITIVE_INFINITY;
+        private double maxY = Double.NEGATIVE_INFINITY;
+        private double minNegX = Double.POSITIVE_INFINITY;
+        private double maxNegX = Double.NEGATIVE_INFINITY;
+        private double minPosX = Double.POSITIVE_INFINITY;
+        private double maxPosX = Double.NEGATIVE_INFINITY;
+
+        private final boolean wrapLongitude;
+
+        public GeoPointVisitor(boolean wrapLongitude) {
+            this.wrapLongitude = wrapLongitude;
+        }
+
+        @Override
+        public void visitPoint(double x, double y) {
+            minY = Math.min(minY, y);
+            maxY = Math.max(maxY, y);
+            visitLongitude(x);
+        }
+
+        @Override
+        public void visitRectangle(double minX, double maxX, double maxY, double minY) {
+            this.minY = Math.min(this.minY, minY);
+            this.maxY = Math.max(this.maxY, maxY);
+            visitLongitude(minX);
+            visitLongitude(maxX);
+        }
+
+        private void visitLongitude(double x) {
+            if (x >= 0) {
+                minPosX = Math.min(minPosX, x);
+                maxPosX = Math.max(maxPosX, x);
+            } else {
+                minNegX = Math.min(minNegX, x);
+                maxNegX = Math.max(maxNegX, x);
+            }
+        }
+
+        @Override
+        public boolean isValid() {
+            return minY != Double.POSITIVE_INFINITY;
+        }
+
+        @Override
+        public Rectangle getResult() {
```

Comment:
To allow this to be used outside of this class, consider extracting this to a static method, maybe in another `class`.